nnef/src/resource.rs

```diff
@@ -0,0 +1,119 @@
+use std::path::Path;
+
+use crate::ast::{Document, QuantFormat};
+use crate::internal::*;
+use tract_core::downcast_rs::{impl_downcast, DowncastSync};
+
+pub const NNEF_DOCUMENT_FILE: &'static str = "graph.nnef";
+pub const QUANT_FORMAT_FILE: &'static str = "graph.quant";
+
+fn path_to_id(path: impl AsRef<Path>) -> TractResult<String> {
+    let mut path = path.as_ref().to_path_buf();
+    path.set_extension("");
+    path.to_str()
+        .ok_or_else(|| format_err!("Badly encoded filename for path: {:?}", path))
+        .map(|s| s.to_string())
+}
+
+pub trait Resource: DowncastSync + std::fmt::Debug + Send + Sync {
+    /// Get value for a given key.
+    fn get(&self, _key: &str) -> TractResult<Value> {
+        bail!("No key access supported by this resource");
+    }
+}
+
+impl_downcast!(sync Resource);
+
+pub trait ResourceLoader: Send + Sync {
+    /// Name of the resource loader.
+    fn name(&self) -> Cow<str>;
+    /// Try to load a resource give a path and its corresponding reader.
+    /// None is returned if the path is not accepted by this loader.
+    fn try_load(
+        &self,
+        path: &Path,
+        reader: &mut dyn std::io::Read,
+    ) -> TractResult<Option<(String, Arc<dyn Resource>)>>;
+
+    fn into_boxed(self) -> Box<dyn ResourceLoader> where Self: Sized + 'static {
+        Box::new(self)
+    }
+}
+
+impl Resource for Document {}
+
+#[derive(Clone, Debug, Copy, PartialEq, Eq, Hash)]
+pub struct NnefDocumentLoader;
```

Comment:
I think we should call this guy `GraphNnefLoader`

Reply:
👍